### ISSUE TITLE
Why does the transductive link prediction task not exclude unseen nodes from testing?

### ISSUE THREAD

**Issue by zhiqiuyuan** [yule-BUAA/DyGLib#18]
Sorry to bother you again!

I noticed in your paper that "This task has two settings: the **transductive** setting aims to predict future links between nodes that are **observed** during training, and the **inductive** setting predicts future links between **unseen** nodes." However, in your code implementation, I observed that the test set for **transductive** link prediction includes **_both_** observed nodes and unseen nodes.

I also reviewed the original code implementation of [TGAT](https://github.com/StatsDLMathsRecomSys/Inductive-representation-learning-on-temporal-graphs) and [TGN](https://github.com/twitter-research/tgn) and found that their test sets in the transductive link prediction task are formed in a similar way to yours (i.e., containing both observed nodes and unseen nodes).

Now I am unsure whether the test set in the transductive link prediction task should exclude or include the unseen nodes.

Could you please tell me **_why you choose to include the unseen nodes in the test set for the transductive link prediction task_**? Thank you very much!

**Comment by Chandler-Q** [comment]
It is worth noting that unobserved nodes in the test set are not used for training. In this case, the authors tested the model for transductive and inductive prediction.

**Comment by KuroginQin** [comment]
I appreciate the wonderful contributions of this repository but I am also confused about the **transductive and inductive settings**. As mentioned by @zhiqiuyuan, I also found that the validation and test sets of transductive settings include **both old observed and new unobserved nodes in the training set**. It seems that this setting is a comprehensive evaluation involving both transductive and inductive settings, which can be a good simulation of real applications. However, it is inconsistent with the corresponding descriptions in the paper.

As I can check, the implementation of [CAW](https://github.com/snap-stanford/CAW) strictly follows the transductive and inductive settings and thus is different from that of this repository. Also as mentioned by @zhiqiuyuan, TGAT and TGN use the same implementation with this repository. 

In summary, it is unclear for me, which setting (i.e., that of CAW or this repository) I should use in my experiments. Many new submissions directly used the results in this paper with the same 'inconsistent' descriptions regarding the inductive and transductive settings.

**Comment by MrYaoH** [comment]
I appreciate the wonderful  contributions of this repository, but the transductive setting makes me confused. I also appreciate the investigation of @KuroginQin, and doubly check the implementation of [CAW](https://github.com/snap-stanford/CAW) for the transductive setting, which seems to use same settings with unseen nodes in the testing (as shown in the "**_arg.mode == 't'_**" condition branch in [main.py](https://github.com/snap-stanford/CAW/blob/master/main.py)), although it employs more detailed and exact inductive settings. 

Notably, I observe that [TAP-GNN](https://github.com/doujiang-zheng/TAP-GNN) strictly uses the seen nodes only in the transductive link prediction, which complies with the corresponding descriptions in the paper and most existing works. Of course, this should be double checked, to make sure the correctness.

I expect that this disagreement could be solved as soon as possible, leading to more robust, unified and precise criterion.

**Comment by QianChang-page** [comment]
I appreciate the wonderful contributions of this repository. Thank @zhiqiuyuan and @MrYaoH for hightligting this issue. I realized that my previous comment was wrong because I didn't understand the key implications.

I suppose we could remove the unobserved nodes from the test set. 

After defining `new_node_val_mask` and `new_node_test_mask`:
https://github.com/yule-BUAA/DyGLib/blob/3aacc36b94b8d2d8293d70a74fdf6d39089b4163/utils/DataLoader.py#L140-L144

Then, we could remove the unobserved nodes from the validation and test sets:
```{python}
    # remove unobserved nodes from the validation and test sets
    val_mask = np.logical_and(val_mask, observed_edges_mask)
    test_mask =  np.logical_and(test_mask, observed_edges_mask)
```

I hope it works.

**Comment by Postroggy** [comment]
Thank @zhiqiuyuan  and @MrYaoH  for hightligting this issue.
If this modification is genuinely necessary, it would render the experimental results of this paper and the model performance reported in the subsequent paper unreliable. This would necessitate retraining the models entirely from scratch. Moreover, the training process is exceedingly lengthy, and increasing the `Batch_Size` is not a viable option to shorten it.
